Improve column auto-width calculations

westonganger opened the pull request:
Solves https://github.com/caxlsx/caxlsx/issues/43

Please review.

noniq commented:
Thanks for the PR! Please have a look into the failing tests (see Travis log) – I suppose there are tests testing the width calculations that need to be adapted for this change.

westonganger commented:
Upon second look it appears this change will cause confusion between input/output. I am instead going to try to remove this code from my library instead.

westonganger commented:
Turns out the original monkey patch was totally flawed.

The problem appears to lie within calculation of autowidth for strings. I have updated the PR accordingly. We shouldnt calculate the autowidth based on any specific font so I have removed the THIN_CHARS constant and it results in a more consistent output.

noniq commented on test/workbook/worksheet/tc_cell.rb:
```diff
@@ -61,7 +61,7 @@ def test_name
     style = @c.row.worksheet.workbook.styles.add_style({:alignment => {:horizontal => :center, :vertical => :center, :wrap_text => true}}  )
     @c.style = style
-    assert_equal(@c.autowidth, 5.5)
+    assert_equal(@c.autowidth, 7.300000000000001)
```
This test fails because the actual value seems to be 3.3 (see https://travis-ci.com/caxlsx/caxlsx/jobs/284684518#L543 )

In general I think it would be preferable to use `#assert_in_delta` here.

noniq commented:
Thanks again for updating the PR (and especially for also thinking about adding to CHANGELOG 😊)

Out of curiosity: Do you have an idea why in the test case the calculated width is now *smaller* than before? From my understanding (and from your comments, too) I got the impression that the cells become rather larger than smaller after this change?

westonganger commented:
Because this line used to add `+ 3` just cause.

https://github.com/caxlsx/caxlsx/pull/44/files#diff-bb22da4127f4d04c73f988f3256690d7L412

noniq commented:
Ah I see, makes sense. So very small cells now get even smaller. Do you think this is a problem? Could cells become too small?

westonganger commented:
Hmm you  are correct. I appear to be observing better results with `+1`

westonganger commented:
- +1 gives a tighter result for small and large cells

![Screenshot from 2020-02-07 16-56-35](https://user-images.githubusercontent.com/3414795/74076138-06d48080-49cb-11ea-8a9c-8e89102816a1.png)

- +3 gives a looser result for small and large cells

![Screenshot from 2020-02-07 16-56-15](https://user-images.githubusercontent.com/3414795/74076164-1fdd3180-49cb-11ea-99fd-7ffd4024cf29.png)

westonganger commented:
+3 may be the best for both small cells and legacy reasons.

westonganger commented:
I think the best is str.size+1 with a minimum final value of 3.

![Screenshot from 2020-02-07 17-18-44](https://user-images.githubusercontent.com/3414795/74076612-fd004c80-49cd-11ea-98a0-557cac4e523c.png)

westonganger commented:
I think that +3 yields the best results and is better for legacy reasons. I have updated the PR accordingly.

westonganger commented on lib/axlsx/workbook/worksheet/rich_text_run.rb:
```diff
@@ -215,7 +215,7 @@ def to_xml_string(str = '')
     def string_width(string, font_size)
       font_scale = font_size / 10.0
-      string.count(Worksheet::THIN_CHARS) * font_scale
+      string.size * font_scale
```
Should this be updated to match the +3 even though the original did not have it?

noniq replied:
Not sure … I’m gonna leave this to your judgement :-)

noniq commented:
Great, I think this is now ready to merge! 

Please add a (very) short summary of the change (and the thought process that went into it) to the commit message before we merge.

westonganger commented:
Ok I have now updated the Changelog to give better explanation/reasoning.

I also changed `RichTextRun` to utilize `Cell#string_width` and `Cell#font_size` instead of re-implementing the exact same methods.

noniq commented:
I think the changes to `RichTextRun` introduce a bug: Since a rich text run can have a style different from the cell’s style, the autowidth calculation needs to take its own styles into account. Here’s a very simple example:

```ruby
require_relative "lib/axlsx"

worksheet = Axlsx::Package.new.workbook.add_worksheet(name: "foo")
rt_1 = Axlsx::RichText.new
rt_1.runs << Axlsx::RichTextRun.new('foobar', sz: 10)
rt_2 = Axlsx::RichText.new
rt_2.runs << Axlsx::RichTextRun.new('foobar', sz: 20)
row = worksheet.add_row [rt_1, rt_2]

puts row[0].autowidth, row[1].autowidth
```

The cells should have different autowidths (because they are using different font sizes), but with the changes applied their widths are identical. (Sadly there seems to be no test case for this exact behaviour).

Let’s backtrack for a moment (I think we’re in danger of getting lost in details): What’s the initial idea behind this PR? Which aspects of autowidth calculation are we trying to fix / improve?

westonganger commented:
Existing behaviour makes cells too small. This PR generally increases cell sizes and provides more predictable autowidth calculation behavior.

As such I have reverted the changes to `RichTextRun` and instead limited those changes to simply include the `+3` in its `string_width` method. The test case that I changed leads me to believe this is correct behaviour because the original test outputs `0`, which no cell should ever autowidth to 0 I dont think.

noniq commented on lib/axlsx/workbook/worksheet/cell.rb:
```diff
@@ -421,6 +416,12 @@ def font_size
+    def styles
+      row.worksheet.styles
+    end
+
```
I think this change is no longer necessary (the `private` modifier and the method can retain their original position, to minimize the diff).

noniq commented on lib/axlsx/workbook/worksheet/rich_text_run.rb:
```diff
@@ -215,7 +215,7 @@ def to_xml_string(str = '')
     def string_width(string, font_size)
       font_scale = font_size / 10.0
-      string.count(Worksheet::THIN_CHARS) * font_scale
+      (string.size + 3) * font_scale
```
`+ 3` here is not necessary: The result of this calculation is used in https://github.com/caxlsx/caxlsx/blob/c30b99b932d06c895b0ca4e14c7b2af6a877cd29/lib/axlsx/workbook/worksheet/cell.rb#L378 
There it gets added to the autowidth calculated for an empty cell, which already contains the fixed offset `3`. So adding the offset here, too, results in it being applied twice.

noniq commented on test/workbook/worksheet/tc_rich_text_run.rb:
```diff
@@ -149,7 +149,7 @@ def test_multiline_autowidth
     awtr.autowidth(ar)
     assert_equal(ar.length, 2)
-    assert_equal(ar.last, 0)
+    assert_in_delta(4.95, ar.last, 0.1)
```
`ar.last` should still be zero: Note that `ar` is an array of values, representing the widths of each line of the text run. Last line is empty, so its width should be zero. However, adding a second assertion testing the calculated value in `ar[0]` would make sense, I suppose.

westonganger commented:
Ok I have addressed your latest concerns.

noniq commented:
Now it’s perfect! Thanks a lot for your dedication to this.
